src/stats/probs/cdf.rs

```diff
@@ -5,6 +5,106 @@
+//! use bio::stats::probs::{LogProb, Prob};
+//! use bio::stats::probs::cdf::{CDF, Entry};
+//! use approx::assert_relative_eq;
+//! // pmf1 is an example PMF with `LogProb(0.0)` at `0`, with `LogProb(0.1)`
+//! // at `{1, 2, ..., 8}` and LogProb(0.2) at `10`
+//! let mut pmf1 = vec![
+//!     Entry::new( 0, LogProb( (0.0 as f64).ln() ) )
+//! ];
+//! for i in 1..=8 {
+//!     pmf1.push(
+//!         Entry::new( i, LogProb( (0.1 as f64).ln() ) )
+//!     );
+//! }
+//! pmf1.push( Entry::new( 10, LogProb( (0.2 as f64).ln() ) ) );
+//!
+//! // create the cumulative distribution function from the probability mass function
+//! let cdf = CDF::from_pmf(pmf1.clone());
+//! assert_relative_eq!(
+//!     *cdf.get(&0).unwrap(),
+//!     (0.0 as f64).ln(),
+//!     epsilon = 0.0
+//! );
+//! assert_relative_eq!(
+//!     *cdf.get(&3).unwrap(),
+//!     (0.3 as f64).ln(),
+//!     epsilon = 0.0000000001
+//! );
+//!
+//! // get back the original probability mass value at 7
+//! assert_relative_eq!(
+//!     *cdf.get_pmf(&7).unwrap(),
+//!     (0.1 as f64).ln(),
+//!     epsilon = 0.00001
+//! );
+//!
+//! // Check that cdf sums up to 1.0
+//! assert_relative_eq!(
+//!     f64::from(cdf.total_prob()),
+//!     (1.0 as f64).ln(),
+//!     epsilon = 0.0
+//! );
+//!
+//! // copy a CDF via its iter() function
+//! let mut cdf_copy = CDF::from_cdf(cdf.iter().cloned());
+//! assert_eq!(cdf.len(), cdf_copy.len());
+//!
+//! // get the maximum a posteriori probability estimate
+//! assert_eq!( cdf_copy.map().unwrap(), &10 );
+//!
+//! // get the 50% credible interval
+//! assert_eq!( cdf_copy.credible_interval(0.5).unwrap(), &2..&8 );
+//!
+//! // cdf_vec is an example Entry vector with `LogProb(0.0)` at `ordered_float::NotNan`
+//! // values `{0.0, 1.0, 2.0}` and increasing by `LogProb(0.2)` at each to `{3.0, 4.0, ..., 7.0}`
+//! let mut cdf_vec = Vec::new();
+//! for i in 0..=2 {
+//!     cdf_vec.push(
+//!         Entry::new(
+//!             NotNan::new(i as f64).unwrap(),
+//!             LogProb::ln_zero()
+//!         )
+//!     )
+//! }
+//! for i in 3..=7 {
+//!     cdf_vec.push(
+//!         Entry::new(
+//!             NotNan::new(i as f64).unwrap(),
+//!             LogProb( ((i-2) as f64 * 0.2f64).ln() )
+//!         )
+//!     );
+//! }
+//!
+//! // create cdf from vector of `Entry`s
+//! let mut cdf_from_vec = CDF::from_cdf(cdf_vec.into_iter());
+//!
+//! assert_relative_eq!(
+//!     *cdf_from_vec.get(&NotNan::new(2.0).unwrap()).unwrap(),
+//!     LogProb::ln_zero(),
+//!     epsilon = 0.0
+//! );
+//! assert_relative_eq!(
+//!     *cdf_from_vec.get(&NotNan::new(4.0).unwrap()).unwrap(),
+//!     LogProb( (0.4 as f64).ln() ),
+//!     epsilon = 0.0
+//! );
+//!
+//! // get the number of `Entry`s in cdf_from_vec
+//! assert_eq!(cdf_from_vec.len(), 8);
+//!
+//! // remove three zero values at `{0.0, 1.0, 2.0}` with `CDF::reduce()`
+//! cdf_from_vec = CDF::reduce(cdf_from_vec);
+//! assert_eq!(cdf_from_vec.len(), 5);
+//! ```
```

Comment:
I'd personally prefer if there were multiple smaller examples; however, having to copy the cdf init code everywhere is probably just as annoying

Reply:
Yeah, I agree. I started out with smaller examples, but the init code got out of hand very quickly. And then I tried with private utility functions right here in `cdf.rs`, but I couldn't get that to work for doctests. In a way, I think this exposes some weaknesses of syntax, e.g. the `LogProb` instantiation / building code is not very readable and maybe the builders for `CDF`s could also be improved. And a formal definition and implementation of a `PMF` is lacking and could further improve syntax. So I guess, the `stats::probs` section in general could do with some systematic additions (and an overhaul to current idiomatic Rust, see e.g. the `CDF::iter()` function), but I guess these are beyond the documentation changes in this PR.
Or we actually go and contribute whatever we have here and doesn't exist over there to the actively growing [statrs crate](https://crates.io/crates/statrs): https://github.com/boxtown/statrs